### ISSUE TITLE
Add custom dataloader

### ISSUE THREAD

**Issue by lishiyu005** [ultralytics/ultralytics#3913]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Dear ultralytics team,

I need to load some extra label file in addition to bboxes. I have changed some function in data/datasets.py

class YOLODataset(BaseDataset):
    def cache_labels(self, path=Path('./labels.cache')):
                    x['labels'].append(
                        dict(
                            im_file=im_file,
                            shape=shape,
                            cls=lb[:, 0:1],  # n, 1
                            bboxes=lb[:, 1:],  # n, 4
                            segments=segments,
                            keypoints=keypoint,
                            _**Customelabels = customlabels**_
                            normalized=True,
                            bbox_format='xywh'))
    def update_labels_info(self, label):
        """custom your label format here."""
        # NOTE: cls is not with bboxes now, classification and semantic segmentation need an independent cls label
        # we can make it also support classification and semantic segmentation by add or remove some dict keys there.
        bboxes = label.pop('bboxes')
        segments = label.pop('segments')
        keypoints = label.pop('keypoints', None)
        **_Customlabels = label.pop('Customlabels')_**
        bbox_format = label.pop('bbox_format')
        normalized = label.pop('normalized')
        label['instances'] = Instances(bboxes, segments, keypoints, pose, kpt3d, bbox_format=bbox_format, normalized=normalized)
        return label

However, my custom data is not loaded into batch. Do you have any idea what I can do to load my custom data into batch?

Thank you very much.

Best regards

**Comment by glenn-jocher** [comment]
@lishiyu005 hi there, 

Thanks for using YOLOv8 and for making your own modifications to load additional label files. Your approach indeed seems to be on the right track.

Upon reviewing your code, it seems like you did well in modifying the `YOLODataset` class to include your custom labels. However, keep in mind that modifying the `cache_labels` method and including the custom labels in the `update_labels_info` function is only the first step. 

You also need to make sure that the data loaders return the custom labels when generating batches. This means you need to adjust your data loading strategy in correspondence with your custom labels, typically in the `__getitem__` method of the `YOLODataset` class.

In your `__getitem__` method, ensure your custom labels are being picked while creating a batch. This function returns the image and associated labels (now including your custom ones) for each item in your dataset. Remember to adjust your batch collating function accordingly as it combines multiple data points into one batch.

Also, it might be necessary for you to update the model, loss function, and the prediction steps to accommodate the new labels that you are introducing.

I hope this guidance is helpful for you. Please proceed with these necessary adjustments and let us know of any further issues you encounter.

Best of luck with your project. 

Best regards,
Glenn Jocher & the Ultralytics team.
